src/clients/BundleDataClient/BundleDataClient.ts

```diff
@@ -1109,72 +1144,174 @@ export class BundleDataClient {
                 this.getRelayHashFromEvent(matchedDeposit) === relayDataHash,
                 "Deposit relay hashes should match."
               );
+              v3RelayHashes[relayDataHash].deposits = [matchedDeposit];
 
-              // slow fill requests for deposits from or to lite chains are considered invalid
-              if (matchedDeposit.fromLiteChain || matchedDeposit.toLiteChain) {
-                return;
-              }
-
-              v3RelayHashes[relayDataHash].deposit = matchedDeposit;
-
-              // Note: we don't need to query for a historical fill at this point because a fill
-              // cannot precede a slow fill request and if the fill came after the slow fill request,
-              // we would have seen it already because we would have processed it in the loop above.
               if (
-                // Input and Output tokens must be equivalent on the deposit for this to be slow filled.
-                !this.clients.hubPoolClient.areTokensEquivalent(
-                  matchedDeposit.inputToken,
-                  matchedDeposit.originChainId,
-                  matchedDeposit.outputToken,
-                  matchedDeposit.destinationChainId,
-                  matchedDeposit.quoteBlockNumber
-                ) ||
+                !_canCreateSlowFillLeaf(matchedDeposit) ||
                 // Deposit must not have expired in this bundle.
-                slowFillRequest.fillDeadline < bundleBlockTimestamps[destinationChainId][1]
+                _depositIsExpired(matchedDeposit)
               ) {
-                // TODO: Invalid slow fill request. Maybe worth logging.
                 return;
               }
               validatedBundleSlowFills.push(matchedDeposit);
             }
           }
         );
 
+        // Deposits can be submitted an arbitrary amount of time after matching fills and slow fill requests.
+        // Therefore, let's go through each deposit in this bundle again and check a few things in order:
+        // - Has the deposit been filled ? If so, then we need to issue a relayer refund for
+        //   this "pre-fill" if the fill took place in a previous bundle.
+        // - Or, has the deposit expired in this bundle? If so, then we need to issue an expiry refund.
+        // - And finally, has the deposit been slow filled? If so, then we need to issue a slow fill leaf
+        //   for this "pre-slow-fill-request" if this request took place in a previous bundle.
+        await mapAsync(bundleDepositHashes, async (depositHash) => {
+          // We don't need to call verifyFillRepayment() here to replace the fill.relayer because this value should already
```

Comment:
I don't think this comment is correct. I think you do need to call `verifyFillRepayment` here since I can have an invalid pre-fill. E.g., I pre-fill in bundle 1 with invalid parameters, I deposit in bundle 2. The pre-fill is never overwritten to the correct refund parameters since it is now outside the bundle block ranges. https://github.com/across-protocol/sdk/blob/f853323fd684b11f4e9bc5efefad2fb5fa77fb9b/src/clients/BundleDataClient/BundleDataClient.ts#L918

Reply:
This is a good point. What do you think of this: 87cb051462af1f0ed36d04f5f6338ccbb7bc8c41 (unit tests are added [here](https://github.com/across-protocol/relayer/pull/2010/commits/65791812f6b407c402aaf3b14c18a43e14f18fa6))